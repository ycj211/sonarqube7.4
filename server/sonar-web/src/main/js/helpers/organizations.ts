/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';
import { Organization, OrganizationSubscription, CurrentUser } from '../app/types';

export function isPaidOrganization(organization: Organization | undefined): boolean {
  return Boolean(organization && organization.subscription === OrganizationSubscription.Paid);
}

export function hasPrivateAccess(
  currentUser: CurrentUser,
  organization: Organization | undefined,
  userOrganizations: Organization[]
): boolean {
  return (
    !isPaidOrganization(organization) ||
    isCurrentUserMemberOf(currentUser, organization, userOrganizations)
  );
}

export function isCurrentUserMemberOf(
  currentUser: CurrentUser,
  organization: Organization | undefined,
  userOrganizations: Organization[]
): boolean {
  return Boolean(
    organization &&
      isLoggedIn(currentUser) &&
      ((organization.actions && organization.actions.admin) ||
        userOrganizations.some(org => org.key === organization.key))
  );
}
