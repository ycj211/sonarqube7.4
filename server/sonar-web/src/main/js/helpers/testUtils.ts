/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow, ShallowRendererProps, ShallowWrapper, ReactWrapper } from 'enzyme';
import { InjectedRouter } from 'react-router';
import { IntlProvider } from 'react-intl';

export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', { ...mockEvent, ...event });
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export function keydown(keyCode: number): void {
  const event = new KeyboardEvent('keydown', { keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.body.clientWidth/clientHeight` are getters by default, so we need to redefine them
  // pass `configurable: true` to allow to redefine the properties multiple times
  if (width) {
    Object.defineProperty(document.body, 'clientWidth', { configurable: true, value: width });
  }
  if (height) {
    Object.defineProperty(document.body, 'clientHeight', { configurable: true, value: height });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

// Create the IntlProvider to retrieve context for wrapping around.
const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();
export function shallowWithIntl<C extends React.Component>(
  node: React.ReactElement<any>,
  options: ShallowRendererProps = {}
) {
  return shallow<C>(node, { ...options, context: { intl, ...options.context } });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}

export function mockRouter(overrides: { push?: Function; replace?: Function } = {}) {
  return {
    createHref: jest.fn(),
    createPath: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    isActive: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    setRouteLeaveHook: jest.fn(),
    ...overrides
  } as InjectedRouter;
}
