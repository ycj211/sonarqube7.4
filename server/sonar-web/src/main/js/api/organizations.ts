/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON } from '../helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import {
  Organization,
  OrganizationBase,
  Paging,
  OrganizationMember,
  Extension
} from '../app/types';

export function getOrganizations(data: {
  organizations?: string;
  member?: boolean;
}): Promise<{
  organizations: Organization[];
  paging: Paging;
}> {
  return getJSON('/api/organizations/search', data).catch(throwGlobalError);
}

export function getOrganization(key: string): Promise<Organization | undefined> {
  return getJSON('/api/organizations/search', { organizations: key }).then(
    r => r.organizations.find((o: Organization) => o.key === key),
    throwGlobalError
  );
}

interface GetOrganizationNavigation {
  adminPages: Extension[];
  canUpdateProjectsVisibilityToPrivate: boolean;
  isDefault: boolean;
  pages: Extension[];
}

export function getOrganizationNavigation(key: string): Promise<GetOrganizationNavigation> {
  return getJSON('/api/navigation/organization', { organization: key }).then(
    r => r.organization,
    throwGlobalError
  );
}

export function createOrganization(
  data: OrganizationBase & { installationId?: string }
): Promise<Organization> {
  return postJSON('/api/organizations/create', data).then(r => r.organization, throwGlobalError);
}

export function updateOrganization(key: string, changes: OrganizationBase) {
  return post('/api/organizations/update', { key, ...changes }).catch(throwGlobalError);
}

export function deleteOrganization(key: string) {
  return post('/api/organizations/delete', { key }).catch(throwGlobalError);
}

export function searchMembers(data: {
  organization?: string;
  p?: number;
  ps?: number;
  q?: string;
  selected?: string;
}): Promise<{ paging: Paging; users: OrganizationMember[] }> {
  return getJSON('/api/organizations/search_members', data).catch(throwGlobalError);
}

export function addMember(data: {
  login: string;
  organization: string;
}): Promise<OrganizationMember> {
  return postJSON('/api/organizations/add_member', data).then(r => r.user, throwGlobalError);
}

export function removeMember(data: { login: string; organization: string }) {
  return post('/api/organizations/remove_member', data).catch(throwGlobalError);
}

export interface OrganizationBilling {
  nclocCount: number;
  subscription: {
    plan?: {
      maxNcloc: number;
      price: number;
    };
    nextBillingDate?: string;
    status: 'active' | 'inactive' | 'suspended';
    trial: boolean;
  };
}

export function getOrganizationBilling(organization: string): Promise<OrganizationBilling> {
  return getJSON('/api/billing/show', { organization, p: 1, ps: 1 });
}
