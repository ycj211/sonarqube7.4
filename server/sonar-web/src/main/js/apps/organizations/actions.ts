/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Dispatch } from 'redux';
import * as api from '../../api/organizations';
import * as actions from '../../store/organizations';
import { addGlobalSuccessMessage } from '../../store/globalMessages';
import { translate, translateWithParameters } from '../../helpers/l10n';
import { Organization, OrganizationBase } from '../../app/types';

export const fetchOrganization = (key: string) => (dispatch: Dispatch) => {
  return Promise.all([api.getOrganization(key), api.getOrganizationNavigation(key)]).then(
    ([organization, navigation]) => {
      if (organization) {
        const organizationWithPermissions = { ...organization, ...navigation };
        dispatch(actions.receiveOrganizations([organizationWithPermissions]));
      }
    }
  );
};

export const createOrganization = (organization: OrganizationBase) => (dispatch: Dispatch<any>) => {
  return api.createOrganization(organization).then((organization: Organization) => {
    dispatch(actions.createOrganization(organization));
    dispatch(
      addGlobalSuccessMessage(translateWithParameters('organization.created', organization.name))
    );
    return organization;
  });
};

export const updateOrganization = (key: string, changes: OrganizationBase) => (
  dispatch: Dispatch<any>
) => {
  return api.updateOrganization(key, changes).then(() => {
    dispatch(actions.updateOrganization(key, changes));
    dispatch(addGlobalSuccessMessage(translate('organization.updated')));
  });
};

export const deleteOrganization = (key: string) => (dispatch: Dispatch<any>) => {
  return api.deleteOrganization(key).then(() => {
    dispatch(actions.deleteOrganization(key));
    dispatch(addGlobalSuccessMessage(translate('organization.deleted')));
  });
};
