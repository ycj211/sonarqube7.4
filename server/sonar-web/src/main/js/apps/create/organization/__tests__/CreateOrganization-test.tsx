/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { times } from 'lodash';
import { Location } from 'history';
import { shallow } from 'enzyme';
import { CreateOrganization } from '../CreateOrganization';
import { mockRouter, waitAndUpdate } from '../../../../helpers/testUtils';
import { LoggedInUser } from '../../../../app/types';
import {
  getAlmAppInfo,
  getAlmOrganization,
  listUnboundApplications
} from '../../../../api/alm-integration';
import { getSubscriptionPlans } from '../../../../api/billing';
import { getOrganizations } from '../../../../api/organizations';
import { get, remove } from '../../../../helpers/storage';

jest.mock('../../../../api/billing', () => ({
  getSubscriptionPlans: jest
    .fn()
    .mockResolvedValue([{ maxNcloc: 100000, price: 10 }, { maxNcloc: 250000, price: 75 }])
}));

jest.mock('../../../../api/alm-integration', () => ({
  getAlmAppInfo: jest.fn().mockResolvedValue({
    application: {
      backgroundColor: 'blue',
      iconPath: 'icon/path',
      installationUrl: 'https://alm.installation.url',
      key: 'github',
      name: 'GitHub'
    }
  }),
  getAlmOrganization: jest.fn().mockResolvedValue({
    almOrganization: {
      avatar: 'my-avatar',
      description: 'Continuous Code Quality',
      key: 'sonarsource',
      name: 'SonarSource',
      personal: false,
      url: 'https://www.sonarsource.com'
    }
  }),
  listUnboundApplications: jest.fn().mockResolvedValue([])
}));

jest.mock('../../../../api/organizations', () => ({
  getOrganizations: jest.fn().mockResolvedValue({ organizations: [] })
}));

jest.mock('../../../../helpers/storage', () => ({
  get: jest.fn().mockReturnValue(undefined),
  remove: jest.fn()
}));

const user: LoggedInUser = {
  groups: [],
  isLoggedIn: true,
  login: 'luke',
  name: 'Skywalker',
  scmAccounts: [],
  showOnboardingTutorial: false
};

beforeEach(() => {
  (getAlmAppInfo as jest.Mock<any>).mockClear();
  (getAlmOrganization as jest.Mock<any>).mockClear();
  (listUnboundApplications as jest.Mock<any>).mockClear();
  (getSubscriptionPlans as jest.Mock<any>).mockClear();
  (getOrganizations as jest.Mock<any>).mockClear();
  (get as jest.Mock<any>).mockClear();
  (remove as jest.Mock<any>).mockClear();
});

it('should render with manual tab displayed', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(getSubscriptionPlans).toHaveBeenCalled();
  expect(getAlmAppInfo).not.toHaveBeenCalled();
});

it('should preselect paid plan on manual creation', async () => {
  const location = { state: { paid: true } };
  // @ts-ignore avoid passing everything from WithRouterProps
  const wrapper = shallowRender({ location });
  await waitAndUpdate(wrapper);
  expect(wrapper.find('ManualOrganizationCreate').prop('onlyPaid')).toBe(true);
});

it('should render with auto tab displayed', async () => {
  const wrapper = shallowRender({ currentUser: { ...user, externalProvider: 'github' } });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(getAlmAppInfo).toHaveBeenCalled();
  expect(listUnboundApplications).toHaveBeenCalled();
});

it('should render with auto tab selected and manual disabled', async () => {
  const wrapper = shallowRender({
    currentUser: { ...user, externalProvider: 'github' },
    location: { query: { installation_id: 'foo' } } as Location
  });
  expect(wrapper).toMatchSnapshot();
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(getAlmAppInfo).toHaveBeenCalled();
  expect(getAlmOrganization).toHaveBeenCalled();
  expect(getOrganizations).toHaveBeenCalled();
});

it('should render with auto personal organization bind page', async () => {
  (getAlmOrganization as jest.Mock<any>).mockResolvedValueOnce({
    almOrganization: {
      key: 'foo',
      name: 'Foo',
      avatar: 'my-avatar',
      personal: true
    }
  });
  const wrapper = shallowRender({
    currentUser: { ...user, externalProvider: 'github', personalOrganization: 'foo' },
    location: { query: { installation_id: 'foo' } } as Location
  });
  expect(wrapper).toMatchSnapshot();
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('should slugify and find a uniq organization key', async () => {
  (getAlmOrganization as jest.Mock<any>).mockResolvedValueOnce({
    almOrganization: {
      avatar: 'https://avatars3.githubusercontent.com/u/37629810?v=4',
      key: 'Foo&Bar',
      name: 'Foo & Bar',
      personal: true
    }
  });
  (getOrganizations as jest.Mock<any>).mockResolvedValueOnce({
    organizations: [{ key: 'foo-and-bar' }, { key: 'foo-and-bar-1' }]
  });
  const wrapper = shallowRender({
    currentUser: { ...user, externalProvider: 'github' },
    location: { query: { installation_id: 'foo' } } as Location
  });
  await waitAndUpdate(wrapper);
  expect(getOrganizations).toHaveBeenCalledWith({
    organizations: ['foo-and-bar', ...times(9, i => `foo-and-bar-${i + 1}`)].join(',')
  });
  expect(wrapper.find('AutoOrganizationCreate').prop('almOrganization')).toMatchObject({
    key: 'foo-and-bar-2'
  });
});

it('should switch tabs', async () => {
  const replace = jest.fn();
  const wrapper = shallowRender({
    currentUser: { ...user, externalProvider: 'github' },
    router: mockRouter({ replace })
  });

  replace.mockImplementation(location => {
    wrapper.setProps({ location }).update();
  });

  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();

  (wrapper.find('Tabs').prop('onChange') as Function)('manual');
  expect(wrapper.find('ManualOrganizationCreate').hasClass('hidden')).toBeFalsy();
  expect(wrapper.find('AutoOrganizationCreate').hasClass('hidden')).toBeTruthy();
  (wrapper.find('Tabs').prop('onChange') as Function)('auto');
  expect(wrapper.find('AutoOrganizationCreate').hasClass('hidden')).toBeFalsy();
  expect(wrapper.find('ManualOrganizationCreate').hasClass('hidden')).toBeTruthy();
});

it('should reload the alm organization when the url query changes', async () => {
  const wrapper = shallowRender({ currentUser: { ...user, externalProvider: 'github' } });
  await waitAndUpdate(wrapper);
  expect(getAlmOrganization).not.toHaveBeenCalled();
  wrapper.setProps({ location: { query: { installation_id: 'foo' } } });
  expect(getAlmOrganization).toHaveBeenCalledWith({ installationId: 'foo' });
  wrapper.setProps({ location: { query: {} } });
  expect(wrapper.state('almOrganization')).toBeUndefined();
  expect(listUnboundApplications).toHaveBeenCalledTimes(2);
});

it('should redirect to organization page after creation', async () => {
  const push = jest.fn();
  const wrapper = shallowRender({ router: mockRouter({ push }) });
  await waitAndUpdate(wrapper);

  wrapper.find('ManualOrganizationCreate').prop<Function>('onOrgCreated')('foo');
  expect(push).toHaveBeenCalledWith({
    pathname: '/organizations/foo',
    state: { justCreated: true }
  });

  (get as jest.Mock<any>).mockReturnValueOnce('0');
  wrapper.find('ManualOrganizationCreate').prop<Function>('onOrgCreated')('foo', false);
  expect(push).toHaveBeenCalledWith({
    pathname: '/organizations/foo',
    state: { justCreated: false }
  });
});

it('should redirect to projects creation page after creation', async () => {
  const push = jest.fn();
  const wrapper = shallowRender({ router: mockRouter({ push }) });
  await waitAndUpdate(wrapper);

  (get as jest.Mock<any>).mockReturnValueOnce(Date.now().toString());
  wrapper.find('ManualOrganizationCreate').prop<Function>('onOrgCreated')('foo');
  expect(get).toHaveBeenCalled();
  expect(remove).toHaveBeenCalled();
  expect(push).toHaveBeenCalledWith({
    pathname: '/projects/create',
    state: { organization: 'foo', tab: 'manual' }
  });

  wrapper.setState({ almOrganization: { key: 'foo', name: 'Foo', avatar: 'my-avatar' } });
  (get as jest.Mock<any>).mockReturnValueOnce(Date.now().toString());
  wrapper.find('ManualOrganizationCreate').prop<Function>('onOrgCreated')('foo');
  expect(push).toHaveBeenCalledWith({
    pathname: '/projects/create',
    state: { organization: 'foo', tab: 'auto' }
  });
});

it('should display AutoOrganizationCreate with already bound organization', async () => {
  (getAlmOrganization as jest.Mock<any>).mockResolvedValueOnce({
    almOrganization: {
      avatar: 'https://avatars3.githubusercontent.com/u/37629810?v=4',
      key: 'Foo&Bar',
      name: 'Foo & Bar',
      personal: true
    },
    boundOrganization: { key: 'foobar', name: 'Foo & Bar' }
  });
  (get as jest.Mock<any>).mockReturnValueOnce(Date.now().toString());
  const push = jest.fn();
  const wrapper = shallowRender({
    currentUser: { ...user, externalProvider: 'github' },
    location: { query: { installation_id: 'foo' } } as Location,
    router: mockRouter({ push })
  });
  await waitAndUpdate(wrapper);
  expect(get).toHaveBeenCalled();
  expect(remove).toHaveBeenCalled();
  expect(getAlmOrganization).toHaveBeenCalledWith({ installationId: 'foo' });
  expect(push).not.toHaveBeenCalled();
  expect(wrapper.find('AutoOrganizationCreate').prop('boundOrganization')).toEqual({
    key: 'foobar',
    name: 'Foo & Bar'
  });
});

it('should redirect to org page when already bound and no binding in progress', async () => {
  (getAlmOrganization as jest.Mock<any>).mockResolvedValueOnce({
    almOrganization: {
      avatar: 'https://avatars3.githubusercontent.com/u/37629810?v=4',
      key: 'Foo&Bar',
      name: 'Foo & Bar',
      personal: true
    },
    boundOrganization: { key: 'foobar', name: 'Foo & Bar' }
  });
  const push = jest.fn();
  const wrapper = shallowRender({
    currentUser: { ...user, externalProvider: 'github' },
    location: { query: { installation_id: 'foo' } } as Location,
    router: mockRouter({ push })
  });
  await waitAndUpdate(wrapper);
  expect(getAlmOrganization).toHaveBeenCalledWith({ installationId: 'foo' });
  expect(push).toHaveBeenCalledWith({ pathname: '/organizations/foobar' });
});

function shallowRender(props: Partial<CreateOrganization['props']> = {}) {
  return shallow(
    <CreateOrganization
      createOrganization={jest.fn()}
      currentUser={user}
      deleteOrganization={jest.fn()}
      // @ts-ignore avoid passing everything from WithRouterProps
      location={{}}
      router={mockRouter()}
      skipOnboarding={jest.fn()}
      updateOrganization={jest.fn()}
      userOrganizations={[
        { actions: { admin: true }, key: 'foo', name: 'Foo' },
        { actions: { admin: true }, alm: { key: 'github', url: '' }, key: 'bar', name: 'Bar' },
        { actions: { admin: false }, key: 'baz', name: 'Baz' }
      ]}
      {...props}
    />
  );
}
