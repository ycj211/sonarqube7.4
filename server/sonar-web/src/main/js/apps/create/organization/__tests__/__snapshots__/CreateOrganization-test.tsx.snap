// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with auto personal organization bind page 1`] = `
<AlmApplicationInstalling
  almKey="github"
/>
`;

exports[`should render with auto personal organization bind page 2`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="onboarding.import_organization.personal.page.header"
    titleTemplate="%s"
  />
  <div
    className="sonarcloud page page-limited"
  >
    <header
      className="page-header"
    >
      <h1
        className="page-title big-spacer-bottom"
      >
        onboarding.import_organization.personal.page.header
      </h1>
    </header>
    <AutoPersonalOrganizationBind
      almApplication={
        Object {
          "backgroundColor": "blue",
          "iconPath": "icon/path",
          "installationUrl": "https://alm.installation.url",
          "key": "github",
          "name": "GitHub",
        }
      }
      almInstallId="foo"
      almOrganization={
        Object {
          "avatar": "my-avatar",
          "key": "foo",
          "name": "Foo",
          "personal": true,
        }
      }
      importPersonalOrg={
        Object {
          "actions": Object {
            "admin": true,
          },
          "key": "foo",
          "name": "Foo",
        }
      }
      onOrgCreated={[Function]}
      updateOrganization={[MockFunction]}
      updateUrlQuery={[Function]}
    />
  </div>
</Fragment>
`;

exports[`should render with auto tab displayed 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="onboarding.create_organization.page.header"
    titleTemplate="%s"
  />
  <div
    className="sonarcloud page page-limited"
  >
    <header
      className="page-header"
    >
      <h1
        className="page-title big-spacer-bottom"
      >
        onboarding.create_organization.page.header
      </h1>
      <p
        className="page-description"
      >
        <FormattedMessage
          defaultMessage="onboarding.create_organization.page.description"
          id="onboarding.create_organization.page.description"
          values={
            Object {
              "break": <br />,
              "more": <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                target="_blank"
                to="/documentation/sonarcloud-pricing/"
              >
                learn_more
              </Link>,
              "price": "billing.price_format.10",
            }
          }
        />
      </p>
    </header>
    <Tabs
      onChange={[Function]}
      selected="auto"
      tabs={
        Array [
          Object {
            "key": "auto",
            "node": "onboarding.import_organization.github",
          },
          Object {
            "key": "manual",
            "node": "onboarding.create_organization.create_manually",
          },
        ]
      }
    />
    <ManualOrganizationCreate
      className="hidden"
      createOrganization={[MockFunction]}
      deleteOrganization={[MockFunction]}
      onOrgCreated={[Function]}
      subscriptionPlans={
        Array [
          Object {
            "maxNcloc": 100000,
            "price": 10,
          },
          Object {
            "maxNcloc": 250000,
            "price": 75,
          },
        ]
      }
    />
    <AutoOrganizationCreate
      almApplication={
        Object {
          "backgroundColor": "blue",
          "iconPath": "icon/path",
          "installationUrl": "https://alm.installation.url",
          "key": "github",
          "name": "GitHub",
        }
      }
      almUnboundApplications={Array []}
      className=""
      createOrganization={[MockFunction]}
      onOrgCreated={[Function]}
      unboundOrganizations={
        Array [
          Object {
            "actions": Object {
              "admin": true,
            },
            "key": "foo",
            "name": "Foo",
          },
        ]
      }
      updateUrlQuery={[Function]}
    />
  </div>
</Fragment>
`;

exports[`should render with auto tab selected and manual disabled 1`] = `
<AlmApplicationInstalling
  almKey="github"
/>
`;

exports[`should render with auto tab selected and manual disabled 2`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="onboarding.create_organization.page.header"
    titleTemplate="%s"
  />
  <div
    className="sonarcloud page page-limited"
  >
    <header
      className="page-header"
    >
      <h1
        className="page-title big-spacer-bottom"
      >
        onboarding.create_organization.page.header
      </h1>
      <p
        className="page-description"
      >
        <FormattedMessage
          defaultMessage="onboarding.create_organization.page.description"
          id="onboarding.create_organization.page.description"
          values={
            Object {
              "break": <br />,
              "more": <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                target="_blank"
                to="/documentation/sonarcloud-pricing/"
              >
                learn_more
              </Link>,
              "price": "billing.price_format.10",
            }
          }
        />
      </p>
    </header>
    <Tabs
      onChange={[Function]}
      selected="auto"
      tabs={
        Array [
          Object {
            "key": "auto",
            "node": "onboarding.import_organization.github",
          },
          Object {
            "key": "manual",
            "node": "onboarding.create_organization.create_manually",
          },
        ]
      }
    />
    <ManualOrganizationCreate
      className="hidden"
      createOrganization={[MockFunction]}
      deleteOrganization={[MockFunction]}
      onOrgCreated={[Function]}
      subscriptionPlans={
        Array [
          Object {
            "maxNcloc": 100000,
            "price": 10,
          },
          Object {
            "maxNcloc": 250000,
            "price": 75,
          },
        ]
      }
    />
    <AutoOrganizationCreate
      almApplication={
        Object {
          "backgroundColor": "blue",
          "iconPath": "icon/path",
          "installationUrl": "https://alm.installation.url",
          "key": "github",
          "name": "GitHub",
        }
      }
      almInstallId="foo"
      almOrganization={
        Object {
          "avatar": "my-avatar",
          "description": "Continuous Code Quality",
          "key": "sonarsource",
          "name": "SonarSource",
          "personal": false,
          "url": "https://www.sonarsource.com",
        }
      }
      almUnboundApplications={Array []}
      className=""
      createOrganization={[MockFunction]}
      onOrgCreated={[Function]}
      unboundOrganizations={
        Array [
          Object {
            "actions": Object {
              "admin": true,
            },
            "key": "foo",
            "name": "Foo",
          },
        ]
      }
      updateUrlQuery={[Function]}
    />
  </div>
</Fragment>
`;

exports[`should render with manual tab displayed 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="onboarding.create_organization.page.header"
    titleTemplate="%s"
  />
  <div
    className="sonarcloud page page-limited"
  >
    <header
      className="page-header"
    >
      <h1
        className="page-title big-spacer-bottom"
      >
        onboarding.create_organization.page.header
      </h1>
      <p
        className="page-description"
      >
        <FormattedMessage
          defaultMessage="onboarding.create_organization.page.description"
          id="onboarding.create_organization.page.description"
          values={
            Object {
              "break": <br />,
              "more": <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                target="_blank"
                to="/documentation/sonarcloud-pricing/"
              >
                learn_more
              </Link>,
              "price": "billing.price_format.10",
            }
          }
        />
      </p>
    </header>
    <ManualOrganizationCreate
      className=""
      createOrganization={[MockFunction]}
      deleteOrganization={[MockFunction]}
      onOrgCreated={[Function]}
      subscriptionPlans={
        Array [
          Object {
            "maxNcloc": 100000,
            "price": 10,
          },
          Object {
            "maxNcloc": 250000,
            "price": 75,
          },
        ]
      }
    />
  </div>
</Fragment>
`;

exports[`should switch tabs 1`] = `
<Fragment>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="onboarding.create_organization.page.header"
    titleTemplate="%s"
  />
  <div
    className="sonarcloud page page-limited"
  >
    <header
      className="page-header"
    >
      <h1
        className="page-title big-spacer-bottom"
      >
        onboarding.create_organization.page.header
      </h1>
      <p
        className="page-description"
      >
        <FormattedMessage
          defaultMessage="onboarding.create_organization.page.description"
          id="onboarding.create_organization.page.description"
          values={
            Object {
              "break": <br />,
              "more": <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                target="_blank"
                to="/documentation/sonarcloud-pricing/"
              >
                learn_more
              </Link>,
              "price": "billing.price_format.10",
            }
          }
        />
      </p>
    </header>
    <Tabs
      onChange={[Function]}
      selected="auto"
      tabs={
        Array [
          Object {
            "key": "auto",
            "node": "onboarding.import_organization.github",
          },
          Object {
            "key": "manual",
            "node": "onboarding.create_organization.create_manually",
          },
        ]
      }
    />
    <ManualOrganizationCreate
      className="hidden"
      createOrganization={[MockFunction]}
      deleteOrganization={[MockFunction]}
      onOrgCreated={[Function]}
      subscriptionPlans={
        Array [
          Object {
            "maxNcloc": 100000,
            "price": 10,
          },
          Object {
            "maxNcloc": 250000,
            "price": 75,
          },
        ]
      }
    />
    <AutoOrganizationCreate
      almApplication={
        Object {
          "backgroundColor": "blue",
          "iconPath": "icon/path",
          "installationUrl": "https://alm.installation.url",
          "key": "github",
          "name": "GitHub",
        }
      }
      almUnboundApplications={Array []}
      className=""
      createOrganization={[MockFunction]}
      onOrgCreated={[Function]}
      unboundOrganizations={
        Array [
          Object {
            "actions": Object {
              "admin": true,
            },
            "key": "foo",
            "name": "Foo",
          },
        ]
      }
      updateUrlQuery={[Function]}
    />
  </div>
</Fragment>
`;
